src/DcmTrajectoryGenerator.cpp

```diff
@@ -0,0 +1,652 @@
+  m_pauseActive(false)
+{}
+
+DcmTrajectoryGenerator::DcmTrajectoryGenerator(const double &dT, const double &omega):
+  m_omega(omega),
+  m_dT(dT),
+  m_pauseActive(false)
+{}
+
+void DcmTrajectoryGenerator::setOmega(const double &omega)
+{
+  m_omega = omega;
+}
+
+void DcmTrajectoryGenerator::setdT(const double &dT)
+{
+  m_dT = dT;
+}
+
+bool DcmTrajectoryGenerator::addLastStep(const double &singleSupportStartTime,
+					 const double &singleSupportEndTime,
+					 const double &singleSupportDuration,
+					 const double &doubleSupportEndTime,
+					 const double &t0,
+					 const iDynTree::Vector2 &comPosition,
+					 const iDynTree::Vector2 &zmp)
+{
+  // evaluate the Single Support trajectory parameters
+  std::shared_ptr<GeneralSupportTrajectory> newSingleSupport = nullptr;
+  newSingleSupport = std::make_shared<SingleSupportTrajectory>(singleSupportStartTime, singleSupportEndTime,
+							       t0, singleSupportDuration, m_omega,
+							       zmp, comPosition);
+  
+  // instantiate position and velocity boundary conditions vectors
+  iDynTree::Vector2 initPosition, initVelocity;
+  iDynTree::Vector2 endPosition, endVelocity;
+
+  // set boundary conditions
+  if(!newSingleSupport->getDcmPos(singleSupportEndTime, initPosition)){
+    std::cerr << "[DCM TRAJECTORY GENERATOR] Error when the position of the DCM in the next SS phase is evaluated." <<std::endl;
+    return false;
+  }
+
+  if(!newSingleSupport->getDcmVel(singleSupportEndTime, initVelocity)){
+    std::cerr << "[DCM TRAJECTORY GENERATOR] Error when the velocity of the DCM in the next SS phase is evaluated." <<std::endl;
+    return false;
+  }
+
+  endPosition = comPosition;
+  endVelocity.zero();
+  double doubleSupportStartTime = singleSupportEndTime;
+  std::shared_ptr<GeneralSupportTrajectory> newDoubleSupport = nullptr;
+  newDoubleSupport = std::make_shared<DoubleSupportTrajectory>(initPosition, initVelocity,
+							       endPosition, endVelocity,
+							       doubleSupportStartTime,
+							       doubleSupportEndTime);
+  // add the new Double Support phase
+  m_trajectory.push_back(newDoubleSupport);
+
+  // add the new Single Support phase
+  m_trajectory.push_back(newSingleSupport);
+  
+  return true;
+}
+
+bool DcmTrajectoryGenerator::addNewStep(const double &singleSupportStartTime,
+					const double &singleSupportEndTime,
+					const double &singleSupportDuration,
+					const double &t0,
+					const iDynTree::Vector2 &zmp)
+{
+  // get the next Single Support trajectory
+  std::shared_ptr<GeneralSupportTrajectory> nextSingleSupport = m_trajectory.back();
+
+  // get the domain of the next SingleSupport trajectory
+  const std::pair<double, double> nextSubTrajectoryDomain = nextSingleSupport->getTrajectoryDomain();
+
+  // get the DCM init of step parameter of the next SingleSupport trajectory
+  std::shared_ptr<SingleSupportTrajectory> nextSingleSupport_cast = std::static_pointer_cast<SingleSupportTrajectory>(nextSingleSupport);
```

Comment:
Why a ``cast`` here?

Reply:
`nextSingleSupport_cast` object is used to get the desired position of the DCM at the beginning of the step [here](https://github.com/GiulioRomualdi/unicycle-footstep-planner/blob/DCM_devel/src/DcmTrajectoryGenerator.cpp#L308) and the position of the ZMP [here](https://github.com/GiulioRomualdi/unicycle-footstep-planner/blob/DCM_devel/src/DcmTrajectoryGenerator.cpp#L358). `cast` is necessary because the base inherited class `GeneralSupportTrajectory` cannot call derived-class methods.

Reply:
So why don't declare it already as a ``SingleSupportTrajectory`` pointer in line 301?

Reply:
Ah no sorry. I see that ``m_trajectory`` contains "generic" pointers.